various changes

bgruening opened the pull request:
Hi @iracooke,

I tried to work on the tests and peptide_prophet seems to work.
Protein Prophet does not have a --database option and the hardcoded path in the XML is obviously wrong. Not sure how this is intended to work. peptide_prophet has a --database option and I have integrated it to make it work.
Interprophet is crashing for me with:

```
Fatal error: Exit code 1 ()
/home/bag/miniconda3/envs/__protk@1.4.4a/lib/ruby/gems/2.4.0/gems/protk-1.4.4/bin/interprophet.rb:81:in `throw': uncaught throw "Command failed with exit code pid 31454 SIGFPE (signal 8) (core dumped)" (UncaughtThrowError)
	from /home/bag/miniconda3/envs/__protk@1.4.4a/lib/ruby/gems/2.4.0/gems/protk-1.4.4/bin/interprophet.rb:81:in `<top (required)>'
	from /home/bag/miniconda3/envs/__protk@1.4.4a/bin/interprophet.rb:23:in `load'
	from /home/bag/miniconda3/envs/__protk@1.4.4a/bin/interprophet.rb:23:in `<main>'
```

I will add a few comments inline. Any help is very appreciated.

bgruening commented on tpp-prophets/peptide_prophet.xml:
```diff
@@ -18,92 +17,66 @@
                 --experiment-label '$experiment_label'
             #end if
-
+            --database '$database'
```
This was needed as otherwise Galaxy was searching for a database table in `/Users/icooke/Desktop/iptest/AASequences.fasta`

bgruening commented on tpp-prophets/peptide_prophet.xml:
```diff
@@ -18,92 +17,66 @@
+    <tests>
+        <test>
+            <param name="input_file" value="mr176-BSA100fmole_BA3_01_8168.d_tandem.pep.xml" ftype="raw_pepxml"/>
+            <param name="database" value="AASequences.fasta" ftype="fasta"/>
```
Without the database it is not working.

bgruening commented on tpp-prophets/protein_prophet.xml:
```diff
@@ -25,11 +22,14 @@
-        <param name="input_files" type="data" format="peptideprophet_pepxml,interprophet_pepxml" multiple="true" label="Peptide Prophet Results" help="These files will typically be outputs from peptide prophet or interprophet"/>
+        <param name="input_files" type="data" format="peptideprophet_pepxml,interprophet_pepxml" multiple="true"
+            label="Peptide Prophet Results" help="These files will typically be outputs from peptide prophet or interprophet"/>
+        <param argument="--database" type="data" format="fasta"
```
Unfortunately, this tool does not seem to have a --database option and crashes because it can not find `/Users/icooke/Desktop/iptest/AASequences.fasta`

iracooke replied:
Yep.  It's tricky to make this work ... but only needed for a test.  Can we find a workaround that only affects testing.  I would prefer to have left the main tool intact

iracooke replied:
I think it's better if this option was left out.  It will most likely confuse users and is unlikely to work in most cases.

bgruening commented on tpp-prophets/pepxml_and_protxml_to_table.xml:
```diff
@@ -0,0 +1,47 @@
+<tool id="protxml_and_pepxml_to_tabular" name="PepXML and ProtXML to Table" version="1.1.1">
+    <description>Converts a pepXML/ProtXML file to a tab delimited text file</description>
```
This tool is a combination of your previous 2 separate tools. depending on the file type is will choose the correct tool to convert.

bgruening commented:
@iracooke progress:

```
proteomics_search_interprophet_1[0]: failed
proteomics_search_peptide_prophet_1[0]: passed
proteomics_search_protein_prophet_1[0]: passed
protxml_and_pepxml_to_tabular[0]: passed
protxml_and_pepxml_to_tabular[1]: passed
```

Interprophet is still crashing, not sure why ... but the others are working now. I have included a ugly hack in ProteinProphet to change the database. See my inline comment.

bgruening commented on tpp-prophets/protein_prophet.xml:
```diff
@@ -1,21 +1,30 @@
+        ln -s '$database' ./AASequences.fasta &&
+        #for $counter, $input_file in enumerate($input_files):
+            ln -s '$input_file' ${counter}.xml &&
+            sed -i.bak 's|<search_database local_path=".*" type="AA"\/>|<search_database local_path="\./AASequences\.fasta" type="AA"\/>|g' ${counter}.xml &&
```
This is my ugly hack to change the XML on the fly and reference the use-given database.

iracooke replied:
Hmm ... this is very clever but I am worried that it will break the tool under normal use.  This should only be needed for tests.  In normal use the database paths will all be set correctly.  My preference would be to have the hack only get run for the test case

iracooke commented on tpp-prophets/interprophet.xml:
```diff
@@ -1,59 +1,31 @@
-            $use_nsm
-            --p-thresh $p_thresh
-            --threads "\${GALAXY_SLOTS:-12}"
+        #for $input_file in $input_files:
```
Are you sure this does the same thing?  The old code was necessary to allow for Galaxy's multifile input support

bgruening replied:
As I understand the code before the only use-case for the old one was that you wanted to take multiple collections and "merge" them, correct? This should be handled by one of the build-in collection tools.

iracooke commented on tpp-prophets/interprophet.xml:
```diff
@@ -1,59 +1,31 @@
-        </conditional>
-
-        <!-- <param name="pepxml_files" multiple="true" type="data" format="peptideprophet_pepxml" label="Peptide Prophet Results" help="These files will typically be outputs from search tools that have subsequently been run through peptide prophet"/> -->
+        <param name="input_files" multiple="true" type="data" format="peptideprophet_pepxml"
```
Ditto comment above. Have you tested with a multifile workflow?  I'm going to trust that you know this much better than I

iracooke commented on tpp-prophets/peptide_prophet.xml:
```diff
@@ -18,92 +17,66 @@
                 --experiment-label '$experiment_label'
             #end if
-
+            --database '$database'
```
This should be an optional argument.  In a normal Galaxy workflow the input to this tool will have the correct path to the database set.  I can see why it would be useful for tests and for unusual workflows though

bgruening replied:
I'm not sure I understand this. The XML file has a hard-coded path to the fasta database and this file is not guaranteed to stay isn't it? Assuming you share this file with me but then clean/delete your history. This file will not work anymore. Its seems this is a very implicit way of encoding a dependency a user is never aware of. Why should a user not delete the fasta file at some point if everything runs "without" this database.

iracooke replied:
@bgruening OK but this terrible behaviour is exactly how the TPP works.  My worry is that there are consequences of this that we have not picked up with our very limited tests.

iracooke commented on tpp-prophets/peptide_prophet.xml:
```diff
@@ -18,92 +17,66 @@
-        <param name="use_only_expect" type="boolean" label="Only use Expect Score as the discriminant" help="Applies only to X!Tandem results. 
+        <param name="input_file" type="data" format="pepxml,raw_pepxml"
+            label="Raw Search Results" help="These files will typically be outputs from omssa or xtandem search tools"/>
+        <param argument="--database" type="data" format="fasta"
```
Please make this optional

iracooke commented on tpp-prophets/peptide_prophet.xml:
```diff
@@ -18,92 +17,66 @@
     </inputs>
     <outputs>
-        <data format="peptideprophet_pepxml" name="output" metadata_source="input_file" label="peptide_prophet.${input_file.display_name}.pep.xml" from_work_dir="peptide_prophet_output.pep.xml"/>
+        <data name="output" from_work_dir="peptide_prophet_output.pep.xml" format="peptideprophet_pepxml"/>
```
Why did you remove the stuff about setting the label? And are you sure it will work now without the from_work_dir?

bgruening replied:
from_work_dir is still there, I just removed the label, but can bring it back. Galaxy has "label" as build in feature nowadays and this might help to archive what you wanted to do here. But I can bring it back.

iracooke replied:
Yes please bring back the label.

bgruening commented:
> Hmm ... this is very clever but I am worried that it will break the tool under normal use. This should only be needed for tests. In normal use the database paths will all be set correctly. My preference would be to have the hack only get run for the test case

We can make the database param hidden and only use it in tests, but I think this is the more correct and explicit way to encode a dependency. Galaxy data management is complex and there are things that will not work with such implicit dependencies. For example consider pulsar/lwr, this allows Galaxy to have not a shared filesystem. Means pulsar will transfer all datasets that are needed for a Job to the compute-node and then executes the job. If the database is not mentioned as part of the job, it will not be transferred, means the job will not find the implicit databse and will crash.

bgruening commented:
@iracooke do you have any idea what is wrong with IProphet?

iracooke commented:
@bgruening no idea sorry

bgruening commented:
@iracooke btw. you can test the tool on usegalaxy.eu if you want. I will fix the labels tomorrow.
Thanks!

bgruening commented:
@iracooke I added the labels back in. I also made the database option optional - I still think this is wrong and will break under many deployments with a very hard to debug error message.
In addition, I included macros and added some other things to make the execution on HPC more robust. Hope that is ok now.

iracooke commented:
Hi @bgruening 
I feel terrible for letting this sit for so long.  To be honest I'm not going to find time to give this a proper test any time soon.  If you are happy that everything works I will merge it.
